fix slice w/ dynamic shapes

cgufb opened the pull request:
Summary: guard_size_oblivious has side effects that'll result in invalid strides when slice nodes take negative index on dynamic input shapes.
Cause overflow error with a huge number “9223372036854776048”
Test Plan: CIs should pass.

Differential Revision: D74354663

laithsakka commented:
do you think this function can be written instead with a combination of guard_or_true, guard_or_false?

we are trying to move away from guard_size_oblivious to more explicit semantics. 

two general approaches come to my mind:
1) is it possible to express the output symbolically with out eager conditioning? is there missing sympy expressions that we would need?
2) shall we assume a more strict semantics and add torch checks?
for example., if start_val is unbacked we assume its positive and add a check  torch.check(start_val>0)
...etc

aside from that if this is blocking you, can you add a unit test that i can debug through so that i can understand the problem you are running into and what you have fixed ?

laithsakka reviewed (changes requested):
please add a unit test

laithsakka commented on torch/_decomp/decompositions.py:
```diff
@@ -739,11 +739,11 @@ def slice_forward(
     elif guard_size_oblivious(start_val > sizes[dim]):
         start_val = sizes[dim]
 
-    if guard_size_oblivious(end_val < start_val):
```
are you saying this is hitting True under unbacked in some situations.

laithsakka commented on torch/_decomp/decompositions.py:
```diff
@@ -739,11 +739,11 @@ def slice_forward(
         start_val = sizes[dim]
 
-    if guard_size_oblivious(end_val < start_val):
+    if statically_known_true(end_val == sys.maxsize):
```
i think the idea here, if end was passed as None then it become sys.maxsize. 
and start was unbacked we do not want end_val = start_val but rather we want, end_val = sizes[dim].

laithsakka commented:
chatting off line this is the graph where this currently fails
```
acc_graph():
    %a : f32[s35, s16][#users=1] = placeholder[target='a']
    %slice_tensor : f32[s35, s16][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %a, dim: 0, start: 0, end: s35, step: 1})
    %slice_tensor_1 : f32[s35, 241][#users=1] = call_function[target=aten.slice.Tensor](kwargs = {input: %slice_tensor, dim: 1, start: Mod(-241, s16), end: s16, step: 1})
    return (%slice_tensor_1,)
```

cgufb committed:
fix slice w/ dynamic shapes (#153131)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/153131

guard_size_oblivious has side effects that'll result in invalid strides when slice nodes take negative index on dynamic input shapes.

Test Plan: CIs should pass.

Differential Revision: D74354663

clee2000 commented:
@pytorchbot merge -f "merged internally"
